Regression(285161@main) Crash in debug under WebSQLiteDatabaseTracker::setIsHoldingLockedFiles(bool)

cdumez opened the pull request:
#### b30afe7d03bc68b7a2928e469e5d6a64352d2515
<pre>
Regression(285161@main) Crash in debug under WebSQLiteDatabaseTracker::setIsHoldingLockedFiles(bool)
<a href="https://bugs.webkit.org/show_bug.cgi?id=281468">https://bugs.webkit.org/show_bug.cgi?id=281468</a>
<a href="https://rdar.apple.com/137923861">rdar://137923861</a>

Reviewed by Ben Nham.

Make sure we call setIsHoldingLockedFiles() on the main run loop since it calls
the `m_isHoldingLockedFilesHandler` lambda which ends up ref&apos;ing the NetworkProcess
object. It is only safe to ref / deref (and use) the NetworkProcess object on the
main run loop.

* Source/WebKit/Shared/WebSQLiteDatabaseTracker.cpp:
(WebKit::WebSQLiteDatabaseTracker::willBeginFirstTransaction):

Canonical link: <a href="https://commits.webkit.org/285177@main">https://commits.webkit.org/285177@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60f0fbd62b7b526d061d895007272da82762a7f1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71766 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51179 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24540 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75881 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22971 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58980 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22791 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56653 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15149 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74832 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46427 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61818 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37107 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43088 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21312 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64994 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19648 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77600 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16000 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/18837 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/64375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16044 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61851 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/64386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15860 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12552 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6188 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46979 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/1758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48050 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49334 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47792 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/Shared/WebSQLiteDatabaseTracker.cpp:
```diff
@@ -69,7 +69,12 @@ void WebSQLiteDatabaseTracker::willBeginFirstTransaction()
-    setIsHoldingLockedFiles(true);
+    ensureOnMainRunLoop([weakThis = WeakPtr { *this }] {
+        if (RefPtr protectedThis = weakThis.get()) {
+            Locker locker { protectedThis->m_lock };
```
Doesn't this deadlock when called on the main thread? The lock is already held starting on line 65.

cdumez replied:
It doesn't happen in practice since we never do database work on the main thread but I've switched to a `RunLoop::main()->dispatch()` call to guarantee it. You're correct that we'd lock recursively otherwise.

bnham reviewed (approved):
LGTM

cdumez committed:
Regression(285161@main) Crash in debug under WebSQLiteDatabaseTracker::setIsHoldingLockedFiles(bool)
https://bugs.webkit.org/show_bug.cgi?id=281468
rdar://137923861

Reviewed by Ben Nham.

Make sure we call setIsHoldingLockedFiles() on the main run loop since it calls
the `m_isHoldingLockedFilesHandler` lambda which ends up ref'ing the NetworkProcess
object. It is only safe to ref / deref (and use) the NetworkProcess object on the
main run loop.

* Source/WebKit/Shared/WebSQLiteDatabaseTracker.cpp:
(WebKit::WebSQLiteDatabaseTracker::willBeginFirstTransaction):

Canonical link: https://commits.webkit.org/285177@main
